### PR TITLE
Refactor scrypt hashing and source app session mgmt pt 3/3

### PR DESCRIPTION
**Pull request by nabla-c0d3**
## Status

Ready.

## Description of Changes

* This PR removes usage of **flask.g** and **flask.session** from the source app's test suite and all of the source app. It's a "cleanup" PR following #5694.
* Update: based on the discussion in #5694, this PR also removes the local cache for source user sessions, in order to fix logout errors/bugs.

### CODE REVIEW

File: securedrop/tests/conftest.py
Comment (nabla-c0d3) [inline securedrop/tests/conftest.py]: This makes the source_user easy to access in the tests

File: securedrop/sdconfig.py
Comment (nabla-c0d3) [inline securedrop/sdconfig.py:92]: I moved the default value from the source app to here.

File: securedrop/source_app/__init__.py
Comment (nabla-c0d3) [inline securedrop/source_app/__init__.py:145]: This was moved to the SessionManager. I think this was the root cause of one of the bugs discussed by @zenmonkeykstop  at the end of #5694 (but not 100% sure).

File: securedrop/source_app/decorators.py
Comment (nabla-c0d3) [inline securedrop/source_app/decorators.py:19]: In this PR, I have removed the local cache (due to the problems encountered in #5694) so a DB session now has to be passed to the SessionManager in order to fetch the source record (instead of fetching it from the now-removed cache).

File: securedrop/source_app/main.py
Comment (nabla-c0d3) [inline securedrop/source_app/main.py:60]: I am not sure why the codenames from the /generate flow need to expire, but this brings the previous behavior back (so that they do expire).

File: securedrop/source_app/session_manager.py
Comment (nabla-c0d3) [inline securedrop/source_app/session_manager.py:30]: I have removed the local cache here to address session timeout bugs discussed in #5694.

File: securedrop/source_app/session_manager.py
Comment (nabla-c0d3) [inline securedrop/source_app/session_manager.py:53]: I've put back the session expiration date in the session cookie; in my previous PR it was in the local cache.

File: securedrop/source_user.py
Comment (nabla-c0d3) [inline securedrop/source_user.py:126]: This should help a lot with performance, by not having to recompute hashes on every authenticated request.

**Review by kushaldas — APPROVED** [review]
Worked well in my staging. I will need a second pair of eyes before merging.

**Comment by eloquence** [comment]
I've done some additional testing with packages built from this branch using the staging env, and so far have not encountered any issues, including by following the steps from https://github.com/freedomofpress/securedrop/pull/5694#issuecomment-916984719 . I'll do additional testing tomorrow 1) with multiple browser instances, 2) with a forced low session expiry setting.

File: securedrop/source_app/main.py
Comment (eloquence) [reply securedrop/source_app/main.py:60]: I don't know if this was the original motivation or not, but one reason I can see to expire codenames is that folks may have left their terminal on the "generate codename" page, then returned to it at a later point, increasing the risk that a third party observed the codename.

**Comment by eloquence** [comment]
I've done additional testing of the following:
- Being logged in as different users in multiple browser instances and logging out / logging back in again. I've observed no issues there.
- Reducing session length expiry to 1 minute and repeatedly bouncing source and journalist sessions, including in multiple tabs. Observed no issues there.
- Letting the session timer expire on the codename page and observing that I get kicked back to the index page. Observed no issues there.

Let me know if you'd like me to test anything else in particular, but so far, the session behavior looks pretty solid in my testing on staging. The behavior also matches what I see in my current 2.0.2 prod environment.

**Comment by zenmonkeykstop** [comment]
Also took a final spin through, via the prod upgrade scenario. Upgrade worked fine, testing with multiple sources worked fine, also loaded a bunch of test data, no issues that I could see.

**Review by conorsch — APPROVED** [review]
Also took this for a spin, and am pleased to say I can't reproduce any of the session-related problems I was experiencing on develop. Also made sure to follow the clear STR in https://github.com/freedomofpress/securedrop/pull/5694#issuecomment-916984719 , and everything worked well. 

Thanks, @zenmonkeykstop, for all your careful review of these changes. And thanks to @nabla-c0d3 as well, for your collaboration throughout review!

**Comment by nabla-c0d3** [comment]
Thanks everyone for taking the time to review and test these changes in depth 😀

**Pull request merged**
